internal/codespaces/portforwarder/port_forwarder.go

```diff
@@ -0,0 +1,244 @@
+	return &PortForwarder{
+		connection: *codespaceConnection,
+	}, nil
+}
+
+// ForwardAndConnectToPort forwards a port and connects to it via a local TCP port.
+func (fwd *PortForwarder) ForwardAndConnectToPort(ctx context.Context, remotePort uint16, listen *net.TCPListener, keepAlive bool, internal bool) error {
+	return fwd.ForwardPort(ctx, remotePort, listen, keepAlive, true, internal, "")
+}
+
+// ForwardPort forwards a port and optionally connects to it via a local TCP port.
+func (fwd *PortForwarder) ForwardPort(ctx context.Context, remotePort uint16, listen *net.TCPListener, keepAlive bool, connect bool, internal bool, visibility string) error {
+	tunnelPort := tunnels.NewTunnelPort(remotePort, "", "", tunnels.TunnelProtocolHttp)
+
+	// If no visibility is provided, Dev Tunnels will use the default (private)
+	if visibility != "" {
+		// Check if the requested visibility is allowed
+		allowed := false
+		for _, allowedVisibility := range fwd.connection.AllowedPortPrivacySettings {
+			if allowedVisibility == visibility {
+				allowed = true
+				break
+			}
+		}
+
+		// If the requested visibility is not allowed, return an error
+		if !allowed {
+			return fmt.Errorf("visibility %s is not allowed", visibility)
+		}
+
+		accessControlEntries := visibilityToAccessControlEntries(visibility)
+		if len(accessControlEntries) > 0 {
+			tunnelPort.AccessControl = &tunnels.TunnelAccessControl{
+				Entries: accessControlEntries,
+			}
+		}
+	}
+
+	// Tag the port as internal or user forwarded so we know if it needs to be shown in the UI
+	if internal {
+		tunnelPort.Tags = []string{InternalPortTag}
+	} else {
+		tunnelPort.Tags = []string{UserForwardedPortTag}
+	}
+
+	// Create the tunnel port
+	_, err := fwd.connection.TunnelManager.CreateTunnelPort(ctx, fwd.connection.Tunnel, tunnelPort, fwd.connection.Options)
+	if err != nil && !strings.Contains(err.Error(), "409") {
+		return fmt.Errorf("create tunnel port failed: %v", err)
+	}
+
+	done := make(chan error)
+	go func() {
+		// Connect to the tunnel
+		err = fwd.connection.TunnelClient.Connect(ctx, "")
+		if err != nil {
+			done <- fmt.Errorf("connect failed: %v", err)
+			return
+		}
+
+		// Inform the host that we've forwarded the port locally
+		err = fwd.connection.TunnelClient.RefreshPorts(ctx)
+		if err != nil {
+			done <- fmt.Errorf("refresh ports failed: %v", err)
+			return
+		}
+
+		// If we don't want to connect to the port, exit early
+		if !connect {
+			done <- nil
+			return
+		}
+
+		// Ensure the port is forwarded before connecting
+		err = fwd.connection.TunnelClient.WaitForForwardedPort(ctx, remotePort)
+		if err != nil {
+			done <- fmt.Errorf("wait for forwarded port failed: %v", err)
+			return
+		}
+
+		// Connect to the forwarded port via a local TCP port
+		err = fwd.connection.TunnelClient.ConnectToForwardedPort(ctx, listen, remotePort)
+		if err != nil {
+			done <- fmt.Errorf("connect to forwarded port failed: %v", err)
+			return
+		}
+	}()
```

Comment:
Do we need a default `done <- nil` here? Otherwise it seems like the `select` at the end of the method will wait until context closes. Is that expected?

Reply:
Yep we definitely should have one there, seems like I missed it (but included it [here](https://github.com/dmgardiner25/cli/blob/648650e1ddc4b86e051d54a4cd2cfd0dff730f4d/internal/codespaces/portforwarder/port_forwarder.go#L175) 🙃). Doesn't really matter though because the `ConnectToForwardedPort` won't return until the user (or context) closes.